tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx

```diff
@@ -49,29 +66,82 @@ private:
+class RSampleSet;
 
-   RDatasetSpec(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
-                const REntryRange &entryRange = {});
+class RSpecBuilder {
+   friend RSampleSet;
 
-   void AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "");
+   std::vector<std::string> fTreeNames;
+   std::vector<std::string> fFileNameGlobs;
+   RDatasetSpec::REntryRange fEntryRange;             // global! range
+   ROOT::TreeUtils::RFriendInfo fFriendInfo;          // global! friends
+   std::vector<RDatasetSpec::RGroupInfo> fGroupInfos; // groups have relative order!
 
-   void
-   AddFriend(const std::string &treeName, const std::vector<std::string> &fileNameGlobs, const std::string &alias = "");
+public:
+   RSpecBuilder &AddGroup(const std::string &groupName, const std::string &treeName, const std::string &fileNameGlob,
+                          const RMetaData &metaData = {});
 
-   void AddFriend(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
-                  const std::string &alias = "");
+   RSpecBuilder &AddGroup(const std::string &groupName, const std::string &treeName,
+                          const std::vector<std::string> &fileNameGlobs, const RMetaData &metaData = {});
 
-   const std::vector<std::string> &GetTreeNames() const;
-   const std::vector<std::string> &GetFileNameGlobs() const;
-   Long64_t GetEntryRangeBegin() const;
-   Long64_t GetEntryRangeEnd() const;
-   const ROOT::TreeUtils::RFriendInfo &GetFriendInfo() const;
+   RSpecBuilder &AddGroup(const std::string &groupName,
+                          const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
+                          const RMetaData &metaData = {});
+
+   RSpecBuilder &AddGroup(const std::string &groupName, const std::vector<std::string> &trees,
+                          const std::vector<std::string> &files, const RMetaData &metaData = {});
+
+   RSpecBuilder &AddGroup(const RSampleSet &sampleSet);
+
+   RSpecBuilder &
+   WithFriends(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "");
+
+   RSpecBuilder &WithFriends(const std::string &treeName, const std::vector<std::string> &fileNameGlobs,
+                             const std::string &alias = "");
+
+   RSpecBuilder &WithFriends(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
+                             const std::string &alias = "");
+
+   RSpecBuilder &WithFriends(const std::vector<std::string> &trees, const std::vector<std::string> &files,
+                             const std::string &alias = "");
+
+   RSpecBuilder &WithRange(const RDatasetSpec::REntryRange &entryRange = {});
+
+   RDatasetSpec Build();
+};
+
+class RSampleSet {
+   std::vector<std::string> fTreeNames;
+   std::vector<std::string> fFileNameGlobs;
+   RDatasetSpec::RGroupInfo fGroupInfo;
```

Comment:
the fact that RSampleSet needs to access RGroupInfo suggests that RGroupInfo shouldn't be nested in RDatasetSpec